### PR TITLE
 [JENKINS-71438] Replace browser confirm with modal dialogs

### PR DESCRIPTION
**Pull request by mawinter69**
# Dialogs
Add a new component for displaying dialogs in javascript as a replacement for the browser built-in dialogs and also have forms in dialogs.

See [JENKINS-71438](https://issues.jenkins.io/browse/JENKINS-71438).

### Supported methods:
- alert
- confirm
- prompt
- modal
- form

### Behaviour
Behaviour is as close as possible to the browser built in functions though there can't be a 1:1 replacement as the dialogs are executed asynchronously.
confirm, prompt and form return a Promise. The success is called when pressing
the OK button, the reject when pressing cancel or escape.

Forms can either submit the page or provide a FormData object to the success case of the Promise.

alert, confirm, prompt and form require to press the buttons, or use the Enter or Escape keys to close the dialog. A modal can be closed by clicking outside of the dialog.

### Usage
See https://github.com/jenkinsci/design-library-plugin/pull/264 for a working example
```
dialog.alert("Title");
dialog.alert("Title", {message: "Message"});

dialog.confirm("Are you sure", {message: "Do it"}).then(
  () => {
    doit();
  },
  ()=>{//This is the cancel case nothing todo
  }
);
  
dialog.prompt("Enter your name:").then(name => {
    console.log("Your name is: " + name);
  }, () => {}
);
```

### Additional changes
- This change also enables storing a message in the context menu, that is shown together with the title when deleting a project (or other places where a confirmation link is used). So instead of just having the browser confirm dialog that tells us we want to delete a project but not which one exactly, we now see the actual project.
- provides a `<l:delete message="" title="" permission="" />`  shortcut for calling the doDelete url on an item as confirmation link. 

### Internationalization
For commonly used texts for buttons a resource bundle is defined that is loaded via javascript.
These texts can also be used by plugins to avoid that they have to provide their own translations.

### Open points
- [x] allow to set the button color in <l:task or <l:confirmationLink
- [x] translation
- [x] form dialogs (with either submitting or returning the data as FormData object)
- [x] link confirmation dialogs
- [x] disable ok button when input of prompt is empty
- [x] Design Library
- [x] confirm used in user token and userseed
- [x] do not close dialog when clicking outside of dialog (=force using the buttons)
- [x] make the methods globally available so they can be easily used in plugins
- [x] need to escape displayname
- [x] bind enter to ok button
- [x] modal

### Where can we use it
- deleting things (done here at many places)
- Disconnecting agents
- Taking agents offline temporarily, update offline reason
- adding things (jobs, views, computers, clouds, users, ...)
- ...

Plugins where dialogs make sense:
- matrix-auth: adding user or group
- role-strategy: adding role (with pattern), adding a user/group
- agent-maintenance: deleting maintenance (single or selected),  adding new maintenance (already uses a dialog but not based on the `dialog` element of html).
- everywhere, where a browser built-in confirm, alert or prompt is used.
- ...

### Ideas
- Icons in the dialog either in the title or the message

### Screenshots:
image

image

image

image

image

### Testing done
Tested interactively the places where confirmation links are used. 
Confirmation links already have a dedicated test, that is passed.
Besides that in the design library change linked above are more demos.

### Proposed changelog entries

- Replace browser confirm with modal dialogs at many places
- Developer: API for alert, confirm, prompt, modal and form dialogs

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] The Jira issue, if it exists, is well-described.
- [x] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)).
  - Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [x] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.

### Desired reviewers

@jenkinsci/sig-ux 

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/7938"></a>

### CODE REVIEW

**Comment by janfaracik** [comment]
Hey, looking good. I've had a WIP for a while now to do this - take a look if you need design ideas etc.

https://github.com/jenkinsci/jenkins/compare/master...janfaracik:jenkins:add-confirmation-modals?expand=1

**Comment by mawinter69** [comment]
Thanks for the info. I'm looking now into making this a general feature so we can replace all confirm, alerts and prompts easily, i.e. making the things blocking so you can just write, also for plugins
```
if (dialog.confirm("Do delete this?")) {
  delete()
}
```#
Still some way to go though

**Comment by mawinter69** [comment]
Damn, the htmlunit doesn't support the showModal method of a dialog element in javascript

**Comment by basil** [comment]
This repository is now using HtmlUnit 3.3.0, which supports modal dialogs, at HEAD.

File: war/src/main/js/components/dialogs/index.js
Comment (mawinter69) [inline war/src/main/js/components/dialogs/index.js:1]: Should this be put into the modals component instead and renamed to window.modal?
The modal/index.js would be now obsolete

File: war/src/main/scss/modules/dialogs.scss
Comment (mawinter69) [inline war/src/main/scss/modules/dialogs.scss:7]: Still uses some definitions from the modals as they are already supported by the black theme.

File: war/src/main/scss/modules/dialogs.scss
Comment (mawinter69) [inline war/src/main/scss/modules/dialogs.scss:8]: defined in modals.scss

File: war/src/main/scss/modules/dialogs.scss
Comment (mawinter69) [inline war/src/main/scss/modules/dialogs.scss:14]: defined in modals.scss

File: war/src/main/scss/modules/dialogs.scss
Comment (mawinter69) [inline war/src/main/scss/modules/dialogs.scss:36]: Buttons are centered in the dialogs. With OK button to the left. Is this ok or should they be put to the left or right side of the dialog and/or have the OK button on the right?

File: war/src/main/scss/modules/dialogs.scss
Comment (mawinter69) [inline war/src/main/scss/modules/dialogs.scss:72]: There are currently no animate-out transitions. When I add them then htmlunit tests are not properly performing the page submit after clicking the ok button. On a live system everything works fine.

**Comment by mawinter69** [comment]
@jenkinsci/sig-ux

File: war/src/main/scss/modules/dialogs.scss
Comment (uhafner) [inline war/src/main/scss/modules/dialogs.scss:3]: I wonder why our dialogs do not have a clear visual separation of the title and the button bar. Everything uses the same color and no line separates these elements. But maybe this is a matter of preference. I prefer the Boostrap styling in this case...

Comment (mawinter69) [reply war/src/main/scss/modules/dialogs.scss:3]: Something like this:
image

or this:
image

File: core/src/main/resources/hudson/model/ManageJenkinsAction/index.properties
Comment (mawinter69) [reply core/src/main/resources/hudson/model/ManageJenkinsAction/index.properties:23]: removed

File: war/src/main/scss/modules/dialogs.scss
Comment (timja) [reply war/src/main/scss/modules/dialogs.scss:3]: I'm not sold on the buttons in the middle currently, 

see:
https://uxdesign.cc/buttons-placement-and-order-bb1c4abadfcb

which recommends on the right

and on button ordering:
https://ux.stackexchange.com/questions/1072/ok-cancel-on-left-right

I think generally in Jenkins we have confirm on the left and cancel on the right (but may be wrong)

Comment (timja) [reply war/src/main/scss/modules/dialogs.scss:3]: (I see you've moved them to the right in later commits than I was testing in the design library pull request)

What do you think about the button order?

Comment (uhafner) [reply war/src/main/scss/modules/dialogs.scss:3]: > Something like this

Yes, I think such a styling will make the different areas more obvious.

Comment (mawinter69) [reply war/src/main/scss/modules/dialogs.scss:3]: What we have on config pages is that buttons are on the left and we have first `Save` then `Apply`
Maybe we should also make the button placement depend on the dialog type.
For `form` dialogs it might be better to have the buttons on the left with `Submit` being the first similar to the configure pages. This might also apply to the `prompt` dialog.
For other dialogs we can put them to the right and try to detect the OS and then put `OK` left (windows) or right (Mac, Linux).

Comment (uhafner) [reply war/src/main/scss/modules/dialogs.scss:3]: I'm not a designer, maybe @janfaracik has an idea what looks better?

Comment (timja) [reply war/src/main/scss/modules/dialogs.scss:3]: > What we have on config pages is that buttons are on the left and we have first `Save` then `Apply` Maybe we should also make the button placement depend on the dialog type. For `form` dialogs it might be better to have the buttons on the left with `Submit` being the first similar to the configure pages. This might also apply to the `prompt` dialog. For other dialogs we can put them to the right and try to detect the OS and then put `OK` left (windows) or right (Mac, Linux).

This sounds good to me.

Comment (mawinter69) [reply war/src/main/scss/modules/dialogs.scss:3]: applied now button placement that way.
Also applied now background color to title and buttons

File: core/src/main/resources/jenkins/security/apitoken/LegacyApiTokenAdministrativeMonitor/resources.js
Comment (timja) [inline core/src/main/resources/jenkins/security/apitoken/LegacyApiTokenAdministrativeMonitor/resources.js:110]: looks like a bad merge
```suggestion
                // TODO simplify when Prototype.js is removed
```

File: war/src/main/scss/modules/dialogs.scss
Comment (timja) [reply war/src/main/scss/modules/dialogs.scss:3]: This doesn't look right on dark theme at least:
image

On light it was pretty much invisible on my mac screen:
image

Would a line work instead?

File: core/src/main/resources/hudson/model/ManageJenkinsAction/index.jelly
Comment (timja) [inline core/src/main/resources/hudson/model/ManageJenkinsAction/index.jelly:67]: When I click this it jumps me to the top of the page which is a bit jarring. There's a flash as it happens.

File: core/src/main/resources/hudson/model/User/sidepanel.jelly
Comment (timja) [inline core/src/main/resources/hudson/model/User/sidepanel.jelly:41]: unsure where the change has happened but commenting here as related, if I click on the delete button from this page http://localhost:8080/jenkins/securityRealm/

I get redirected to:
http://localhost:8087/jenkins/securityRealm/user/john/delete

which is a 404 now:
image

File: core/src/main/resources/lib/layout/delete.jelly
Comment (timja) [inline core/src/main/resources/lib/layout/delete.jelly:28]: from what I can tell it's done like this:
```suggestion
    Adds a delete link in the left navigation bar of Jenkins.
    
    @since TODO
```

**Review by timja — COMMENTED** [review]
Looking really good I've tested it quite thoroughly locally.

some issues I found

File: core/src/main/resources/hudson/model/User/sidepanel.jelly
Comment (mawinter69) [reply core/src/main/resources/hudson/model/User/sidepanel.jelly:41]: That was a bug that was fixed with https://github.com/jenkinsci/jenkins/pull/8131

File: core/src/main/resources/hudson/model/ManageJenkinsAction/index.jelly
Comment (mawinter69) [reply core/src/main/resources/hudson/model/ManageJenkinsAction/index.jelly:67]: I think the reason is that confirmationlink renders an `<a href="#" ...`
This makes the browser jump to the top. I think I've seen this behaviour even before.

Comment (mawinter69) [reply core/src/main/resources/hudson/model/ManageJenkinsAction/index.jelly:67]: https://stackoverflow.com/questions/2084750/javascript-anchor-avoid-scroll-to-top-on-click

File: war/src/main/scss/modules/dialogs.scss
Comment (timja) [reply war/src/main/scss/modules/dialogs.scss:3]: It feels weird the buttons moving all over the place depending on the dialog type, it may just be a problem in the design library as you see them all soon after each other and e.g. when you use the prompt one and alert pops up after but the buttons are on different side.

what do you think?

-----

Dark theme is looking good to me although the line seems a bit prominent on the default theme:
image

image

Comment (mawinter69) [reply war/src/main/scss/modules/dialogs.scss:3]: Actually there might be cases where you can see multiple dialogs with different type directly after each other. E.g. in matrix auth, plugin when you add a new user (uses a prompt) and that user is already added you would get an alert.

Comment (mawinter69) [reply war/src/main/scss/modules/dialogs.scss:3]: default theme with `var(--medium-grey)`

image

**Comment by Kevin-CB** [comment]
The security review is currently in progress.

**Review by Kevin-CB — APPROVED** [review]
I have tested your PR locally and from a security perspective, I don't see any blockers.

**Comment by timja** [comment]
ATH is failing https://github.com/jenkinsci/acceptance-test-harness/pull/1256

**Comment by janfaracik** [comment]
Looking great!

Couple of things - 

I'm not sure if the borders benefit the dialog? If there was a lot of content I could see it being useful but right now I think it divides the dialog unnecessarily. 

I'd also make everything a touch chunkier, currently theres varying font sizes and fairly tight padding - making everything a little larger makes it feel more consistent and also makes the hit targets larger, e.g.

image

The code for the above is 

<details>
<summary>Click to expand</summary>
$jenkins-dialog-padding: 1.3rem;

.jenkins-dialog {
  border-radius: 1rem;
  border: none;
  background: transparent;
  box-shadow: var(--dialog-box-shadow);
  backdrop-filter: var(--dialog-backdrop-filter);
  animation: jenkins-dialog-animate-in 0.25s cubic-bezier(0, 0.68, 0.5, 1.5);
  overflow: hidden;
  padding: 0;

  &::backdrop {
    background: var(--dialog-backdrop-background);
    animation: jenkins-dialog-backdrop-animate-in 0.15s;
  }

  &__contents {
    overflow-y: auto;
    overflow-wrap: break-word;
    padding: 0 $jenkins-dialog-padding;
    max-height: 75vh;
    font-size: 1rem;
    color: var(--text-color-secondary);
  }

  &__input {
    display: flex;
    padding: 0 1rem 1rem;
  }

  &__buttons {
    padding: $jenkins-dialog-padding;

    .jenkins-button {
      font-size: 0.9rem;
      padding: 0.7rem 0.6rem;
      min-width: 100px;
    }

    &--right-windows {
      justify-content: right;
    }

    &--right-unix {
      justify-content: right;
      flex-direction: row-reverse;
    }
  }

  &__title {
    font-size: 1.1rem;
    font-weight: 600;
    padding: $jenkins-dialog-padding;
  }

  &__subtitle {
    font-size: 1rem;
    font-weight: 600;
    color: var(--text-color-secondary);
    padding: 0;
    margin: 0 0 1rem;
  }

  &__close-button {
    position: absolute;
    top: $jenkins-dialog-padding - 0.5rem;
    right: $jenkins-dialog-padding - 0.5rem;
    aspect-ratio: 1;
    padding: 0;

    &::before,
    &::after {
      border-radius: 100%;
      outline: none;
    }

    &::after {
      backdrop-filter: contrast(0.9) blur(5px);
    }
  }
}

@keyframes jenkins-dialog-backdrop-animate-in {
  from {
    opacity: 0;
  }
}

@keyframes jenkins-dialog-animate-in {
  from {
    scale: 85%;
    opacity: 0;
  }
}
</details>

---

Also unsure about reordering the buttons depending on OS? I might be alone here but I use Jenkins across a variety of operating systems so having the controls change order would be a little confusing for muscle memory. It also looks like the flex box `flex-direction` doesn't alt + tab correctly.

**Comment by mawinter69** [comment]
Making the message text with secondary color has the drawback that this doesn't look so good when there is no title. One could make the color depending if a title is present or not.
Which button layout would you propose?
- Buttons on the left (OK button left)
- Buttons on the right windows style (OK button left)
- Buttons on the right unixstyle (OK button right)

**Comment by mawinter69** [comment]
> I'd also make everything a touch chunkier, currently theres varying font sizes and fairly tight padding - making everything a little larger makes it feel more consistent and also makes the hit targets larger, e.g.
> 

The proposed change has the problem that when there is no title that then the padding is missing, also for prompts it is not working, the text is directly above the input without any padding then. Probably I need to make the padding depend on title is there or not

**Comment by janfaracik** [comment]
> Making the message text with secondary color has the drawback that this doesn't look so good when there is no title. 

Is there ever a situation where the dialog would have a description but not a title?

> Which button layout would you propose?

I'd go for 'Buttons on the right unixstyle (OK button right)' personally

> also for prompts it is not working, the text is directly above the input without any padding then. Probably I need to make the padding depend on title is there or not

If you made the dialog a flex box, you could use the `gap` attribute to set a gap between each element, you wouldn't need the padding then :)

**Comment by mawinter69** [comment]
> > Making the message text with secondary color has the drawback that this doesn't look so good when there is no title.
> 
> Is there ever a situation where the dialog would have a description but not a title?

I would not exclude that possibility, it is not always necessary to have both.
Instead of having the first parameter be the message I could make it the title for alert, confirm and prompt
> 
> > Which button layout would you propose?
> 
> I'd go for 'Buttons on the right unixstyle (OK button right)' personally

Maybe we need to do a poll. 
As the OK button has different color it already draws attention to it.
> 
> > also for prompts it is not working, the text is directly above the input without any padding then. Probably I need to make the padding depend on title is there or not
> 
> If you made the dialog a flex box, you could use the `gap` attribute to set a gap between each element, you wouldn't need the padding then :)

will try that out

**Comment by mawinter69** [comment]
@janfaracik I'm using now a flex box, made it a bit easier

buttons are now always unix style

The first parameter to alert, confirm and prompt will now be the title and not the message.

Adjusted some occurences of confirms to have both title and message

**Review by timja — APPROVED** [review]
Looks great!

**Comment by janfaracik** [comment]
Looking great! Couple non-blockers from me -

1. There isn't a closing animation (there is for the 'Icon legend' modal)
2. A higher min width would look good, it's a little too narrow in the below example 

image

Something like 450px makes it look better:

image

**Review by janfaracik — APPROVED** [review]

**Comment by mawinter69** [comment]
> 2. A higher min width would look good, it's a little too narrow in the below example

increased minwidth to 450px

**Comment by timja** [comment]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Comment by MarkEWaite** [comment]
Congratulations on the merge of the pull request.  I experimented with it shortly after its release and like the result very much.  It provides a much better user experience than the previous browser delete dialog.

I need your help with the plugin bill of materials to adapt to this pull request.  The Jenkins plugin bill of materials is showing new test failures in the config-file-provider-plugin with Jenkins 2.415.  When I bisect the test failures in Jenkins core, the first commit that shows the test failure is this one.

The test failures are listed in the plugin BOM test results at https://github.com/jenkinsci/bom/runs/15255225821 and are available on ci.jenkins.io at https://ci.jenkins.io/job/Tools/job/bom/job/master/1899/testReport/

@mawinter69 would you be willing to review the 4 test failures and their source code and suggest changes to the config file provider plugin to fix the issue?

To see the issue, I ran the following maven command with a checkout of the master branch of the config file provider plugin:

```
mvn clean -DforkCount=1 -Djenkins.version=2.415 -Dtest=ConfigFilesSEC1253Test.xssPrevention,ConfigFilesSEC1253Test#regularCaseStillWorking,FolderConfigFileActionTest#xssPreventionInFolder,Security2002Test#xssPrevention verify
```

To confirm that 2.414 did not have the issue, I ran the command:

```
mvn clean -DforkCount=1 -Djenkins.version=2.414 -Dtest=ConfigFilesSEC1253Test.xssPrevention,ConfigFilesSEC1253Test#regularCaseStillWorking,FolderConfigFileActionTest#xssPreventionInFolder,Security2002Test#xssPrevention verify
```

**Comment by MarkEWaite** [comment]
@mawinter69 it looks like this change has introduced a regression in the pop-up menu for "Manage Jenkins" that is available when clicking the down arrow of the "Dashboard" link at the top of the page.  More details are described in [JENKINS-71744](https://issues.jenkins.io/browse/JENKINS-71744).

Would you be willing to review it and see what can be done to resolve it?

**Comment by mawinter69** [comment]
I'll take a look

**Pull request merged**
